bcl2fastq/lib/bcl2fastq_utils.py

```diff
@@ -134,32 +140,49 @@ def get_bases_mask_per_lane_from_samplesheet(samplesheet, index_lengths):
         :param samplesheet: samplesheet to fetch the index lengths from
         :param index_lengths: dict of index lengths (e.g. "{1: 7, 2: 8}"),
         normally parsed from run meta data.
+        :param is_single_read: True if this is a single read run, else false.
         :return a dict of the lane and base mask to use, e.g.:
-                 { 1:"y*,iiiiiiiin*,iiiiiiiin*,y*" , 2:"y*,iiiiii,n*,y*  [etc] }
+                 { 1:"y*,i7n*,i7n*,y*" , 2:"y*,i5,n*,y*  [etc] }
         """
 
-        def pad_with_ignore(length_of_index_in_samplesheet, length_of_index_read):
+        def build_index_string(length_tuple):
+            """
+            Builds the index mask string
+            :param length_tuple: a tuple of the length of the index in the samplesheet and in the read, e.g. (3, 5)
+            :return: a index string, e.g. "i5n*" or "n*" or "i3", depending on the situation.
+            """
+            length_of_index_in_samplesheet = length_tuple[0]
+            length_of_index_read = length_tuple[1]
             difference = length_of_index_read - length_of_index_in_samplesheet
+
             assert difference >= 0, "Sample sheet indicates that index is longer than what was read by the sequencer!"
-            if difference > 0:
+
+            if length_of_index_in_samplesheet == 0:
+                # If there is no index in the samplesheet, ignore it in the base-mask
                 return "n*"
+
+            if difference > 0:
+                # Pad the end if there is a difference here.
+                return "i" + str(length_of_index_in_samplesheet) + "n*"
             else:
-                return ""
-
-        def construct_double_index_basemask(index1, index2):
-            index1_length = len(index1)
-            index2_length = len(index2)
-            return "y*,{0}{1}{2},{3}{4}{5},y*".format(
-                "i", index1_length, pad_with_ignore(index1_length, index_lengths[2]),
-                "i", index2_length, pad_with_ignore(index2_length, index_lengths[3]))
-
-        def construct_single_index_basemask(idx, flowcell_has_double_idx):
-            idx_length = len(idx)
-            if flowcell_has_double_idx:
-                return "y*,{0}{1}{2},{3},y*".format(
-                    "i", idx_length, pad_with_ignore(idx_length, index_lengths[2]), "n*")
+                return "i" + str(length_of_index_in_samplesheet)
+
+        def construct_base_mask(samplesheet_idx_list):
+            """
+            Will construct the base mask.
+            :param samplesheet_idx_list: A list of the indexes in the samplesheet
+            :return a base mask string
+            """
+            samplesheet_idx_list = map(len, samplesheet_idx_list)
+            samplesheet_idx_and_read_length_tuples = zip(samplesheet_idx_list, index_lengths.itervalues())
```

Comment:
I like this because it's general - so if you think it's alright I'd like to keep it.